### PR TITLE
Make On*Removed conventions run even if the declaring type was removed previously

### LINKED ISSUE #19311: On*Removed conventions should run even if the declaring type was removed previously
This is especially important for foreign keys as the principal entity type might still be in the model.

Also add debug checks between execution to ensure previous convention didn't change any required state without terminating.

### LINKED ISSUE #20220: Problem with inherited owned type
I have a problem with mapping of a complex domain model.
The problem is the generated mapping of Owned type.
When I created a migration step, then there will be two compiler error in the code.

### Model what I want to map

``` C#
    public class Entity
    {
        public int Id { get; set; }
        public BookingInterval Interval { get; set; }
        public DateRange Range { get; set; }
    }

    public class DateRange
    {
        public DateTime From { get; set; }
        public DateTime To { get; set; }
    }

    public class BookingInterval : DateRange
    {
        public DateTime CheckIn { get; set; }
        public DateTime CheckOut { get; set; }
    }
```

### The Context file with mapping
``` C#
    public class POCContext : DbContext
    {
        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            var entityBuilder = modelBuilder.Entity<Entity>();
            entityBuilder.OwnsOne(p => p.Interval, ob =>
            {
                ob.Property(p => p.CheckIn).IsRequired();
                ob.Property(p => p.CheckOut).IsRequired();
                ob.Ignore(p => p.From);
                ob.Ignore(p => p.To);
            });
            entityBuilder.Ignore(p => p.Range);
            modelBuilder.Ignore<DateRange>();
            base.OnModelCreating(modelBuilder);
        }
    }
```

### The problem
When I call the Add-Migration command, then the following Snapshot will be generated.
``` C#
    [DbContext(typeof(POCContext))]
    partial class POCContextModelSnapshot : ModelSnapshot
    {
        protected override void BuildModel(ModelBuilder modelBuilder)
        {
#pragma warning disable 612, 618
            modelBuilder
                .HasAnnotation("ProductVersion", "3.1.2")
                .HasAnnotation("Relational:MaxIdentifierLength", 128)
                .HasAnnotation("SqlServer:ValueGenerationStrategy", SqlServerValueGenerationStrategy.IdentityColumn);

            modelBuilder.Entity("EFCoreMigrationProblem1.Model.Entity", b =>
                {
                    b.Property<int>("Id")
                        .ValueGeneratedOnAdd()
                        .HasColumnType("int")
                        .HasAnnotation("SqlServer:ValueGenerationStrategy", SqlServerValueGenerationStrategy.IdentityColumn);

                    b.HasKey("Id");

                    b.ToTable("Entity");
                });

            modelBuilder.Entity("EFCoreMigrationProblem1.Model.Entity", b =>
                {
                    b.OwnsOne("EFCoreMigrationProblem1.Model.BookingInterval", "Interval", b1 =>
                        {
                            b1.Property<int>("EntityId")
                                .ValueGeneratedOnAdd()
                                .HasColumnType("int")
                                .HasAnnotation("SqlServer:ValueGenerationStrategy", SqlServerValueGenerationStrategy.IdentityColumn);

                            b1.Property<DateTime>("CheckIn")
                                .HasColumnType("datetime2");

                            b1.Property<DateTime>("CheckOut")
                                .HasColumnType("datetime2");

                            b1.HasKey("EntityId");

                            b1.ToTable("Entity");

                            // The problem is here!
                            b1.HasDiscriminator();

                            b1.WithOwner()
                                .HasForeignKey("EntityId");
                        });
                });
#pragma warning restore 612, 618
        }
    }
```
So the problem is that, the `b1.HasDiscriminator();` method case compiler error.
And same problem is in Designer file of migration step.

### Workaround
I found a workaround to this problem, but it is not fitting well with my domain model.
The model of workaround:
``` C#
    public class Entity
    {
        public int Id { get; set; }
        public BookingInterval Interval { get; set; }
        public DateRange Range { get; set; }
    }
    public class DateRangeBase
    {
        public DateTime From { get; set; }
        public DateTime To { get; set; }
    }
    public class DateRange : DateRangeBase
    {
        public DateTime From { get; set; }
        public DateTime To { get; set; }
    }
    public class BookingInterval : DateRangeBase
    {
        public DateTime CheckIn { get; set; }
        public DateTime CheckOut { get; set; }
    }
```

### Further technical details

EF Core version: 3.1.2
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: All traget
Operating system: Windows 10
IDE: Visual Studio 2019 16.4.5

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fixes #19311
Fixes #20220

### CODE REVIEW

File: src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs [hunk 1]
```diff
@@ -230,14 +232,25 @@ protected virtual void TryConfigurePrimaryKey(IConventionEntityTypeBuilder entit
...
+        }
+    }
 
     /// <inheritdoc />
     public virtual void ProcessKeyRemoved(
         IConventionEntityTypeBuilder entityTypeBuilder,
         IConventionKey key,
         IConventionContext<IConventionKey> context)
     {
+        if (!entityTypeBuilder.Metadata.IsInModel)
```
Comment (ajcvickers) [inline src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs:249]: So is there a general rule for when `IsInModel` should be used?

**Review by ajcvickers — APPROVED** [review]

File: src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs:249]: Always, in the OnRemoved conventions. And elsewhere if you are paranoid

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs:249]: Of course, one won't need to do these checks once we have layering

Comment (ajcvickers) [reply src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs:249]: Of course.

**Pull request merged**